mapBlat
---

Originally designed to map Genomic and RNA reads onto MaxDiversity (MD) regions found on contigs generated by the pAss pipeline.
mapBlat will now be officially integrated into the omics, MetamapsDB ecosystem.

parser scripts


## Usage



generate node and rel CSVs

links to 


### Contig Taxonomic annotation

| Step | Description                                 | Output   |
| 1    | `Diamond blastx` contigs against NR-protein | DAA file |
| 2    |


```
$ ./tools/blast2lca -i <DAA file> -f DAA -o <output file> -g2t /export2/home/uesu/simulation_fr_the_beginning/data/classifier/gi2taxid.refseq
```

#### NOTE 

* has to be executed inside `/path/to/megan/rootDir/`
* use `xvfb-run` if it complatins about not finding a display

For an example for the pipeline use see example.R in the `R/` folder.

[Still under construction]

Required Files:

* Contigs from metagenomic reads belonging to KEGG ortholog
* FASTQ files taken from

